src/ChannelServer/World/SpawnZone.cs

```diff
@@ -0,0 +1,219 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Melia.Shared.Data.Database;
+using Melia.Shared.Util;
+using Melia.Shared.World;
+
+namespace Melia.Channel.World
+{
+
+
+	public struct SpawnData
+	{
+		public string spawnName;
+		public string mapName;
+		public string monsterName;
+		public int count;
+		public int countVariation;
+		public bool isFixedLocation;
+		public Position Position { get; set; }
+		public Direction Direction { get; set; }
+		public int respawnTime;
+		public int xMin, xMax, zMin, zMax;
+	}
+
+	public class SpawnZone
+	{
+
+		public int Id;
+		public SpawnData Data { get; set; }
+		public MonsterData MonsterData { get; set; }
+		private Map _map;
+		private Dictionary<int, Monster> _spawnEntities;
+		private int _calculatedTotalEntities;
+		private int _countEntities;
+		private int _countDeaths;
+		private bool _isInitialized;
+		private bool _respawnEnabled;
+		private Random rndGenerator;
+
+		public SpawnZone(SpawnData spawnData)
+		{
+			this.Data = spawnData;
+			this.MonsterData = ChannelServer.Instance.Data.MonsterDb.Find(this.Data.monsterName);
+
+			if (this.MonsterData == null)
+			{
+				Log.Error("Error initializing Spawn Zone {1} {2}. Monster not found. {0}", this.Data.monsterName, this.Id, this.Data.spawnName);
+				return;
+			}
+
+			_map = ChannelServer.Instance.World.GetMap(this.Data.mapName);
+			if (_map == null)
+			{
+				Log.Error("Error initializing Spawn Zone {1} {2}. Map not found. {0}", this.Data.mapName, this.Id, this.Data.spawnName);
+				return;
+			}
+			Log.Debug("map {0}", _map.Name);
+
+			_isInitialized = false;
+			_countDeaths = 0;
+
+			_spawnEntities = new Dictionary<int, Monster>();
+			rndGenerator = new Random();
+		}
+
+		public void StartRespawn()
+		{
+			if (!_isInitialized)
+			{
+				Log.Warning("Spawn {0} {1} can't start respawn because its not initialized.", this.Id, this.Data.spawnName);
+				return;
+
+			}
+
+			_respawnEnabled = true;
+		}
+
+		public void StopRespawn()
+		{
+			_respawnEnabled = false;
+		}
+
+		public void Init()
+		{
+			if (_map == null || this.MonsterData == null)
+				return;
+
+			// Calculate count of spawn entities
+			Random rnd = new Random();
+			_calculatedTotalEntities = this.Data.count + rnd.Next(this.Data.countVariation);
```

Comment:
You basically always generate a random number to spawn for the entire session? I know monsters spawn counts seem random, but from my experience they vary based on the amount of players, or rather kills in the area. The more monsters are killed in quick succession the more monsters spawn, to account for the increase in player numbers.

An implementation of that would probably use a base count and increase the respawn number by a certain amount, maybe double, based on the kill frequency.